### PR TITLE
Clarification around usage of Union

### PR DESCRIPTION
**Pull request by somada141**
## Change Summary

Adds examples of the incorrect and correct usage of `Union` when defining type annotations with multiple types and updated documentation. This is discussed further under https://github.com/samuelcolvin/pydantic/issues/514

## Checklist

* [ ] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [ ] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

**Comment by somada141** [comment]
@samuelcolvin please have a look when you get a chance

File: docs/index.rst [hunk 1]
```diff
@@ -949,6 +949,24 @@ Benchmarks were run with python 3.7.2 and the following package versions:
...
+Usage of ``Union`` in Annotations and Type Order
+------------------------------------------------
+
+The ``Union`` type allows a model attribute to accept different models (classes), e.g.:
+
+.. literalinclude:: examples/union_type_incorrect.py
+
+(This script is complete, it should run "as is")
+
+However, as can be seen above, *pydantic* will attempt to 'match' any of the models (classes) defined under ``Union`` and will use the first one that matches. In the above example the ``id`` of ``user_03`` was defined as a ``uuid.UUID`` class (which is defined under the attribute's ``Union`` annotation) but as the ``uuid.UUID`` can be marshalled into an ``int`` it chose to match against the ``int`` type and disregarded the other types.
```
Comment (samuelcolvin) [inline docs/index.rst]: please hard wrap this text.

Also `models (classes)` can changed to `types`.

File: docs/index.rst [hunk 2]
```diff
@@ -949,6 +949,24 @@ Benchmarks were run with python 3.7.2 and the following package versions:
...
+
+As such, it is recommended that when defining ``Union`` annotations that the most specific model (class) is defined first and followed by less specific models (classes). In the above example, the ``UUID`` class should precede the ``int`` and ``str`` classes to preclude the unexpected representation as such:
```
Comment (samuelcolvin) [inline docs/index.rst]: same.

File: docs/index.rst [hunk 3]
```diff
@@ -949,6 +949,24 @@ Benchmarks were run with python 3.7.2 and the following package versions:
 * **django-restful-framework** ``v3.9.4``
 
 
+Usage of ``Union`` in Annotations and Type Order
```
Comment (samuelcolvin) [inline docs/index.rst]: can you move this above benchmarks.

File: docs/examples/union_type_correct.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from pydantic import BaseModel
+
+
+class User(BaseModel):
+    id: Union[UUID, int, str]
+    name: str
+
+
+user_03_uuid = UUID('cf57432e-809e-4353-adbd-9d5c0d733868')
+user_03 = User(**{'id': user_03_uuid, 'name': 'John Doe'})
```
Comment (samuelcolvin) [inline docs/examples/union_type_correct.py]:
```suggestion
user_03 = User(id=user_03_uuid, name='John Doe')
```

File: docs/examples/union_type_incorrect.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+from typing import Union
+from pydantic import BaseModel
+
+
+class User(BaseModel):
+    id: Union[int, str, UUID]
+    name: str
+
+
+user_01 = User(**{'id': 123, 'name': 'John Doe'})
```
Comment (samuelcolvin) [inline docs/examples/union_type_incorrect.py]:
```suggestion
user_01 = User(id=123, name='John Doe')
```

File: docs/examples/union_type_incorrect.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+print(user_01)
+# > User id=123 name='John Doe'
+print(user_01.id)
+# > 123
+
+user_02 = User(**{'id': '1234', 'name': 'John Doe'})
```
Comment (samuelcolvin) [inline docs/examples/union_type_incorrect.py]:
```suggestion
user_02 = User(id='1234', name='John Doe')
```

File: docs/examples/union_type_incorrect.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+print(user_02)
+# > User id=1234 name='John Doe'
+print(user_02.id)
+# > 1234
+
+user_03_uuid = UUID('cf57432e-809e-4353-adbd-9d5c0d733868')
+user_03 = User(**{'id': user_03_uuid, 'name': 'John Doe'})
```
Comment (samuelcolvin) [inline docs/examples/union_type_incorrect.py]:
```suggestion
user_03 = User(id=user_03_uuid, name='John Doe')
```

File: docs/index.rst (hunk 3 above)
Comment (somada141) [reply docs/index.rst]: done

File: docs/index.rst (hunk 2 above)
Comment (somada141) [reply docs/index.rst]: done

File: docs/index.rst (hunk 1 above)
Comment (somada141) [reply docs/index.rst]: done

**Comment by somada141** [comment]
@samuelcolvin I updated `HISTORY.rst` but didn't set a version for the release, does the way I added to it make sense?

**Comment by somada141** [comment]
nvm realised you pushed a new version on it. updated

File: HISTORY.rst [hunk 8]
```diff
@@ -6,6 +6,8 @@ History
 v0.29 (unreleased)
 ..................
 * support dataclasses.InitVar, #592 by @pfrederiks
+* Updated documentation to elucidate the usage of ``Union`` when defining multiple types under an attribute's
+  annotation and showcase how the type-order can affect marshalling of provided values.
```
Comment (samuelcolvin) [inline HISTORY.rst]:
```suggestion
  annotation and showcase how the type-order can affect marshalling of provided values, #594 by @somada141
```

File: docs/index.rst [hunk 9]
```diff
@@ -922,6 +922,28 @@ to properly set types before the model can be used.
 
    Resolving this is beyond the call for *pydantic*: either remove the future import or declare the types globally.
 
+Usage of ``Union`` in Annotations and Type Order
+------------------------------------------------
+
+The ``Union`` type allows a model attribute to accept different types, e.g.:
+
+.. literalinclude:: examples/union_type_incorrect.py
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
**(This script is complete, it should run but may be is wrong, see below)**

.. literalinclude:: examples/union_type_incorrect.py
```

I'm bad at reading docs, I search, find the first example that looks write and go with it. Hopefully this message will help other people as lazy as me.

File: docs/index.rst [hunk 10]
```diff
@@ -922,6 +922,28 @@ to properly set types before the model can be used.
...
+
+(This script is complete, it should run "as is")
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
```

File: docs/index.rst [hunk 11]
```diff
@@ -922,6 +922,28 @@ to properly set types before the model can be used.
...
+
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
```

File: docs/index.rst (hunk 9 above)
Comment (somada141) [reply docs/index.rst]: makes sense

**Comment by somada141** [comment]
@samuelcolvin updated

**Comment by samuelcolvin** [comment]
merged, thanks so much.

Docs will be updated on the next release.

**Comment by somada141** [comment]
beautiful thanks a lot for the help @samuelcolvin

**Pull request merged**